### PR TITLE
Additional testcases

### PR DESCRIPTION
**Pull request by jkikstra**
Based on [this suggestion](https://github.com/iiasa/aneris/pull/15#issuecomment-550173406) by @gidden in #15, three additional test cases are introduced in this PR.

Note: the test_data for the regression tests has not been quality-checked/validated (@gidden this is for you to do as an expert on how the harmonization should behave). No overrides are yet provided.

Issue: currently, with the binary switch `global_harmonization_only`, only a subset of the emissions gets harmonized (i.e. only global aggregates or no global aggregates) -- this needs a fix.

### CODE REVIEW

**Comment by gidden** [comment]
@jkikstra I rebased this onto current master so we only see the commits related to this PR

**Comment by gidden** [comment]
Hey @jkikstra this PR now contains passing tests for all cases you provided - I had to do a bit of revision to the test cases, but the basic idea is as follows: for each trajectory defined at region and sector level, apply a simple to compute override (`constant_ratio` and `constant_offset`). This guarantees that the aggregation also works as expected.

I additionally added features for:

- returning directly the objects, see https://github.com/iiasa/aneris/pull/18/files#diff-3437ae6214d4c2f163341d9e5e0853f5R43

- providing a custom suffix, see: https://github.com/iiasa/aneris/pull/18/files#diff-9c03adcf85f191a2de5ff1c6792223f4R4

I think this is the barebones you need to complete the AR6 setup. Once I find infinite time, these things should be better documented, but, sadly, that day is not today.

Let's catch up again this week to confirm that you have everything you need here.

**Comment by gidden** [comment]
thanks @jkikstra - ignore the stickler issues for now, its the first time this repo is being checked this way

**Comment by gidden** [comment]
Ok @jkikstra this PR now holds only the edits you're working on at the moment so we can more easily identify changes!

File: aneris/cli.py [hunk 1]
```diff
@@ -63,25 +64,28 @@ def harmonize(inf, history, regions, rc, output_path, output_prefix, return_resu
...
-    fname = os.path.join(output_path, '{}_metadata.xlsx'.format(prefix))
-    logger().info('Writing metadata to: {}'.format(fname))
-    aneris.pd_write(metadata, fname)
-
-    # save data about harmonization
-    if not diagnostics.empty:
-        fname = os.path.join(output_path, '{}_diagnostics.xlsx'.format(prefix))
-        logger().info('Writing diagnostics to: {}'.format(fname))
-        aneris.pd_write(diagnostics, fname)
+    if write_output:
```
Comment (gidden) [inline aneris/cli.py]: out of curiosity - why is `return_result` not sufficient? if you use that option, then the workflow stops on return?

File: aneris/utils.py [hunk 2]
```diff
@@ -212,7 +213,7 @@ def subtract_regions_from_world(df, name=None, base_year='2015', threshold=5e-2)
         threshold below which to set values to 0
     """
     # make global only global (not global + sum of regions)
-    check_null(df, name)
+    # check_null(df, name)
```
Comment (gidden) [inline aneris/utils.py]: why does this check need to be removed?

Comment (jkikstra) [reply aneris/utils.py]: see comment in commit https://github.com/iiasa/aneris/pull/18/commits/6c1d85c5b75ce47ad63964a7a80d341925daa3d9

>@gidden commenting `check_null(df, name)` solve the KeyError here (see previous commit with global_ar6 test). However, I'm not fully sure what exactly the nature of the problem was and why this line was implemented in the first place. Is commeting this out reasonable?

File: aneris/cli.py (hunk 1 above)
Comment (jkikstra) [reply aneris/cli.py]: Ah my intention was to make the option for writing out the results more explicit and allow for a combination of both options, but now I see I should have changed the order of the two if-statements to make this work.

**Comment by gidden** [comment]
hey @jkikstra - turns out `check_null()` was in fact a canary in the coal mine here. Here's what was happening:

- `check_null()` (yes, poorly named perhaps) checks if any null values exist and drop them
- the drop happens by default row-wise (`axis=0`)
- null values (i.e., missing data) existed in the datasets - notably 2005/2010 data for certain F-gases
- therefore, `check_null()` dropped those rows
- therefore, it appeared like the model was actually missing those data (if you look at your test file, they all have constant values, and if you look at the metadata, it says `model_zero`)

I have updated this so that the drop happens *column wise*. This is probably better for our use case, *but*, if a team present results with missing data for one gas in a column, that whole column would go away.

I think it is probably ok to leave this as is for now, but perhaps it is worth discussing with Ed how to deal with missing data *before* sending it to `aneris`

**Comment by gidden** [comment]
Note please that the test file will need to be updated with correct expected values. Once that is done, we can merge the whole chain.

File: aneris/cli.py (hunk 1 above)
Comment (gidden) [reply aneris/cli.py]: Cool, thanks!

File: aneris/utils.py (hunk 2 above)
Comment (gidden) [reply aneris/utils.py]: See new comments for updated reasoning here

**Comment by jkikstra** [comment]
Great, thanks as lot @gidden for finding the issue here! 

That's a good point you raise. For what we have encountered now: 
- deleting rows (gases) when the missing data is _before the harmonization year_ is unwanted and unnecessary, deleting a column (year) would not be a problem

In other cases:
- however, a missing value _in the harmonization year_ should probably lead to deleting the row (the GHG), as (in my understanding -- or is there interpolation that could be done?) the gas currently could not be harmonized anyway, and deleting the row would lead to no harmonization taking place for any gas.
- with missing values _after the harmonization year_, it is a question of design, and I do not know what would be preferred. Deleting the column would only make sense if the model does not report any gases in these years (which is a likely case, e.g. models reporting only up till 2050). Deleting rows makes more sense if only pathways fully reporting up till 2100 should be taken into account. If we would move away from `check_null()`, other options would open up, such as building in a check that just harmonizes up until the year that has been reported for a certain gas.

**Comment by jkikstra** [comment]
AFAIK @gidden this PR can be merged.

**Comment by jkikstra** [comment]
Hi @gidden, I'll go ahead and merge this as I've now updated the test file and everything passes.

**Pull request merged**